File: app/src/main/java/com/osrsd/cache/provider/VarbitProvider.java
package com.osrsd.cache.provider;

import com.osrsd.cache.def.VarbitDefinition;
import com.osrsd.spring.domain.Definition;
import lombok.extern.slf4j.Slf4j;

import java.nio.ByteBuffer;

@Slf4j
public final class VarbitProvider {

    public static Definition decode(ByteBuffer buffer, VarbitDefinition definition) {
        do {
            int opcode = buffer.get() & 0xff;
            if (opcode == 0) {
                break;
            }

            if (opcode == 1) {
                definition.setIndex(buffer.getShort() & 0xffff);
                definition.setLeastSignificantBit(buffer.get() & 0xff);
                definition.setMostSignificantBit(buffer.get() & 0xff);
            } else {
                log.warn(String.format("Unhandled definition opcode with id: %s.", opcode));
            }
        } while (true);
        return definition;
    }

}

File: app/src/main/kotlin/com/osedit/cache/provider/VarbitProvider.kt
package com.osedit.cache.provider

import com.osedit.spring.domain.Definition
import com.osedit.spring.domain.ItemDefinition
import com.osedit.spring.domain.VarbitDefinition
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer

object VarbitProvider : Provider<VarbitDefinition> {

    private val log: Logger = LoggerFactory.getLogger(VarbitProvider::class.java)

    override fun decode(buffer: ByteBuffer, definition: VarbitDefinition): Definition {
        do {
            when (val opcode: Int = buffer.get().toInt() and 0xff) {
                1 -> {
                    definition.index = buffer.short.toInt() and 0xffff
                    definition.leastSignificantBit = buffer.get().toInt() and 0xff
                    definition.mostSignificantBit = buffer.get().toInt() and 0xff
                }
                0 -> break
                else -> log.warn(String.format("Unhandled definition opcode with id: %s.", opcode))
            }
        } while (true)
        return definition
    }
}
